drivers/bm/bm_dfu_core.c

```diff
@@ -0,0 +1,342 @@
+}
+
+static void s_idle_run(void *o)
+{
+    if (_dfu_context.current_event.type == DFU_EVENT_RECEIVED_UPDATE_REQUEST)
+    {
+#ifdef CONFIG_BM_DFU_CLIENT
+        /* Client */
+        bm_dfu_client_process_request();
+#endif
+    }
+    else if (_dfu_context.current_event.type == DFU_EVENT_BEGIN_UPDATE)
+    {
+#ifdef CONFIG_BM_DFU_HOST
+        /* Host */
+        smf_set_state(SMF_CTX(&_dfu_context), &dfu_states[BM_DFU_STATE_HOST_REQ_UPDATE]);
+#endif
+    }
+}
+
+static void s_error_entry(void *o)
+{
+    /* TODO: What do we do here? */
+    LOG_ERR("Entered the Error State.");
+
+    switch (_dfu_context.error)
+    {
+        case BM_DFU_ERR_FLASH_ACCESS:
+            LOG_ERR("Flash access error (Fatal Error)");
+            break;
+        case BM_DFU_ERR_IMG_CHUNK_ACCESS:
+            LOG_ERR("Unable to get image chunk");
+        case BM_DFU_ERR_NONE:
+        case BM_DFU_ERR_TOO_LARGE:
+            LOG_ERR("Image too large for Client");
+        case BM_DFU_ERR_SAME_VER:
+            LOG_ERR("Client already loaded with image");
+        case BM_DFU_ERR_MISMATCH_LEN:
+            LOG_ERR("Length mismatch");
+        case BM_DFU_ERR_BAD_CRC:
+            LOG_ERR("CRC mismatch");
+        case BM_DFU_ERR_TIMEOUT:
+            LOG_ERR("DFU Timeout");
+        case BM_DFU_ERR_BM_FRAME:
+            LOG_ERR("BM Processing Error");
+        default:
+            smf_set_state(SMF_CTX(&_dfu_context), &dfu_states[BM_DFU_STATE_IDLE]);
+            break;
+    }
+}
+
+static void s_error_exit(void *o)
+{
+    /* TODO: What do we do here? */
+    LOG_ERR("Exited the Error State.");
+}
+
+/* Populate state table */
+static const struct smf_state dfu_states[] = 
+{
+        [BM_DFU_STATE_INIT] = SMF_CREATE_STATE(NULL, s_init_run, NULL, NULL),
+        [BM_DFU_STATE_IDLE] = SMF_CREATE_STATE(NULL, s_idle_run, NULL, NULL),
+        [BM_DFU_STATE_ERROR] = SMF_CREATE_STATE(s_error_entry, NULL, s_error_exit, NULL),
+
+#ifdef CONFIG_BM_DFU_CLIENT
+        /* Child states of BM_DFU_STATE_CLIENT */
+        [BM_DFU_STATE_CLIENT] = SMF_CREATE_STATE(s_client_entry, NULL, s_client_exit, NULL),
+        [BM_DFU_STATE_CLIENT_RECEIVING] = SMF_CREATE_STATE(s_client_receiving_entry, s_client_receiving_run, NULL, &dfu_states[BM_DFU_STATE_CLIENT]),
+        [BM_DFU_STATE_CLIENT_VALIDATING] = SMF_CREATE_STATE(s_client_validating_entry, NULL, NULL, &dfu_states[BM_DFU_STATE_CLIENT]),
+        [BM_DFU_STATE_CLIENT_ACTIVATING] = SMF_CREATE_STATE(s_client_activating_entry, NULL, NULL, &dfu_states[BM_DFU_STATE_CLIENT]),
+#endif
+#ifdef  CONFIG_BM_DFU_HOST
+        /* Host states of BM_DFU_STATE_HOST */
+        [BM_DFU_STATE_HOST] = SMF_CREATE_STATE(s_host_entry, NULL, s_host_exit, NULL),
+        [BM_DFU_STATE_HOST_REQ_UPDATE] = SMF_CREATE_STATE(s_host_req_update_entry, s_host_req_update_run, NULL, &dfu_states[BM_DFU_STATE_HOST]),
+        [BM_DFU_STATE_HOST_UPDATE] = SMF_CREATE_STATE(s_host_update_entry, s_host_update_run, NULL, &dfu_states[BM_DFU_STATE_HOST]),
+#endif
```

Comment:
😱
I'd never seen this type of array initialization! Learn something new every day :D